### PR TITLE
Add support to Typescript's node types

### LINKED ISSUE #71: Consider handling typescript-eslint-specific AST
The team I work on is migrating a Flow project to Typescript, and we are trying to keep our ESLint setup by using [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint), but things were not working out as expected.

After some debugging, we found out that the issue was within `jsx-ast-utils`, which didn't correctly handle JSX attributes in the form of `atribute={value!}` (where `!` is a Typescript non-null assertion, i.e. it tells Typescript to assume that `value` is neither `null` or `undefined`). It crashed [here](https://github.com/evcohen/jsx-ast-utils/blob/master/src/values/expressions/index.js#L112-L114), more specifically. The issue is that `@typescript-eslint/typescript-estree` exposes it as a `TSNonNullExpression` node with `value`'s AST as `node.expression`.

For our codebase, simply adding
```js
if (type === "TSNonNullExpression") {
  expression = expression.expression;
  type = expression.type;
}
```
before that check made everything work. It's a fairly large React codebase, so I guess it's fair to assume there would be little additional work.

I would be happy to come up with a PR if the project sees this as desirable. Thanks!

### PR DESCRIPTION
**Pull request by vhfmag**
This PR:

- Updates babel and related libraries to `^7` for Typescript support
- Updates Jest to ^21 because updating babel broke tests
- Replaces `babylon` with `@babel/parser`for Typescript support
- Implements support for Typescript's `TSNonNullExpression` and `TSAsExpression` node type and tests them

Fixes #71

### CODE REVIEW

**Comment by vhfmag** [comment]
@ljharb Babel 7 dropped support for Node < 6. Maybe this should involve a major version bump?

**Comment by ljharb** [comment]
@vhfmag it'd be better if we could instead change CI to always run babel in latest node LTS, and then switch to the desired node version for running tests.

**Comment by vhfmag** [comment]
Ok! Maybe we should test with latest and LTS?

**Comment by vhfmag** [comment]
@ljharb How would I do this? Generate an AST with `@babel/parser`, save it to a JSON and use it in the tests? I don't know how feasible is this, since there is not a single point where the AST is generated, but rather each test dynamically call the parser with some code.

**Comment by ljharb** [comment]
oh no, not at all. I mean, `nvm install --lts && npm run build && nvm use $actualVersion` in travis.yml

**Comment by vhfmag** [comment]
Oh, understood. I tried transpiling both src and __tests__ folders with `["@babel/preset-env", { "targets": { "node": "4" } }]` in Node 8 and running the transpiled tests with Node 4, but it still did not work. Probably something with `@babel/parser`, which is used in the tests. If retro-compatibility is a hard goal, maybe a bundling `@babel/parser` for Node 4 could solve it. Worth trying?

**Comment by ljharb** [comment]
Babel 7 doesn't support node < 6; I don't think bundling. is going to be particularly great. It'd be nice if the workaround only lived in travis.yml.

**Comment by vhfmag** [comment]
I just don't see how to make it possible to run tests with Node 4, they fail because of the incompatibility of Node 4 :disappointed:

**Comment by ljharb** [comment]
ah - if jest 21 is the one that dropped support for node 4, then we'd have to stick to jest 20.

The reality is that tape is the only test runner that reliably maintains support for older nodes, so eventually anything published will need to switch to it, but that's out of scope for this PR.

**Comment by vhfmag** [comment]
I just checked and Jest 21 supports Node 4 just fine. I meant that `@babel/parser`, used in the tests, does not support it. Maybe I could use either `babylon` or `@babel/parser`, according to which version of node is running.

**Comment by ljharb** [comment]
ahhhhh gotcha. hmm. if there's a way to run the babel 6 parser in node 4, that'd be great - just as you say

**Comment by vhfmag** [comment]
@ljharb done! Everything is working again for Node >= 4. :smile:

File: __tests__/helper.js [hunk 1]
```diff
@@ -1,11 +1,31 @@
+/* globals beforeEach */
```
Comment (ljharb) [inline __tests__/helper.js]:
```suggestion
/* eslint-env jest */
```

**Comment by kripod** [comment]
I just [submitted an issue to typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/issues/437#event-2284964883) and found out this pull request would resolve it. Are there any reasons against merging and releasing this TS-related improvement?

**Comment by ljharb** [comment]
It needs to be rebased.

**Comment by vhfmag** [comment]
@ljharb done! I'll pinpoint the changes I've needed to do to make this easier to review.

File: __tests__/src/getPropValue-test.js [hunk 2]
```diff
@@ -357,17 +359,21 @@ describe('getPropValue', () => {
     });
 
     it('should evaluate to a correct representation of member expression with a nullable member', () => {
-      // This tell will not throw when Babel is upgraded from 6 to 7. Remove
-      // the throw expectation wrapper at that time.
-      // eslint-disable-next-line no-undef
-      expect(() => {
+      const runTest = () => {
```
Comment (vhfmag) [inline __tests__/src/getPropValue-test.js]: Here I have extracted the expected-to-fail function into a variable to conditionally expect it to fail or not according to Babel version

**Comment by kripod** [comment]
I would like to know whether it is possible to issue a new package release soon. This is a dealbreaker for TypeScript developers.

**Comment by evcohen** [comment]
Working on it now @kripod

**Comment by evcohen** [comment]
published [v2.1.0](https://github.com/evcohen/jsx-ast-utils/releases/tag/v2.1.0)

**Comment by kripod** [comment]
Unfortunately, I'm still encountering the issue: https://github.com/typescript-eslint/typescript-eslint/issues/437

```
$ eslint "**/*.{ts,tsx,js}"
Error: The prop value with an expression type of TSNonNullExpression could not be resolved. Please file issue to get this fixed immediately.
Occurred while linting C:\Development\Projects\gatsby-starter-strict\src\components\Layout.tsx:32
    at Object.extractLiteral [as JSXExpressionContainer] (C:\Development\Projects\gatsby-starter-strict\node_modules\jsx-ast-utils\lib\values\expressions\index.js:219:11)
    at getLiteralValue (C:\Development\Projects\gatsby-starter-strict\node_modules\jsx-ast-utils\lib\values\index.js:61:35)
    at extractValue (C:\Development\Projects\gatsby-starter-strict\node_modules\jsx-ast-utils\lib\getPropValue.js:24:12)
    at getLiteralPropValue (C:\Development\Projects\gatsby-starter-strict\node_modules\jsx-ast-utils\lib\getPropValue.js:56:10)
    at JSXAttribute (C:\Development\Projects\gatsby-starter-strict\node_modules\eslint-plugin-jsx-a11y\lib\rules\lang.js:47:58)
    at listeners.(anonymous function).forEach.listener (C:\Development\Projects\gatsby-starter-strict\node_modules\eslint\lib\util\safe-emitter.js:45:58)
    at Array.forEach (<anonymous>)
    at Object.emit (C:\Development\Projects\gatsby-starter-strict\node_modules\eslint\lib\util\safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (C:\Development\Projects\gatsby-starter-strict\node_modules\eslint\lib\util\node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (C:\Development\Projects\gatsby-starter-strict\node_modules\eslint\lib\util\node-event-generator.js:280:22)
```

**Comment by vhfmag** [comment]
Weird, this seems to be precisely the bug this patch was meant to fix. Mind to share the code that causes it?

**Comment by kripod** [comment]
@vhfmag I linked my issue with code included in my previous comment.

**Comment by vhfmag** [comment]
The issue seems to be that `LITERAL_TYPES` in https://github.com/evcohen/jsx-ast-utils/blob/master/src/values/expressions/index.js#L88-L126 does not include an extractor for TSNonNullExpression. I'm planning to work on it soon, but I'm not sure what is expected to be extracted.

Adding `TSNonNullExpression: noop` fixes the lint error, and adding `TSNonNullExpression: value => value.expression.type` doesn't throw but makes for a lint failure in https://github.com/kripod/gatsby-starter-strict.

@evcohen would you explain what is the expected behavior in this case? Thanks in advance!

**Comment by vhfmag** [comment]
@kripod https://github.com/evcohen/jsx-ast-utils/pull/80

**Pull request merged**